stdlib/enum.ncl

```diff
@@ -1,7 +1,7 @@
 {
-    enum = {
-      Tag
-        | doc m%"
+  enum = {
```

Comment:
It seems sometimes there is a newline inserted before the opening `{`, and sometimes not (as here). Is there any rule around this  behavior?

Reply:
It depends on the context. Usually a new line isn't inserted before a record, but there can be cases where a new line is inserted when the following node is multi-line (regardless of it being a record). This happens in `let...in...` expressions, for example:

```nickel
# Single-line
let x = { foo = 123 } in x

# Multi-line
let x =
  {
    foo = 123
  }
in x
```

The fact that the multi-line node is a record, in this case, is circumstantial. It _may_ be possible to add more nuance to these rules...

Reply:
But here, `enum` is multiline too, isn't it? In particular, I'm still wondering why

```
let enum = {

}
in

```

and

```
{
  enum = {
  ...
  },
}
```

would be treated differently, as both bind a name to a record, and both are multi-line. It seems, for some reason, that the first one get a newline before the `{`, but not the second one. Unless the record field isn't multiline, because what is taken into account is the declaration (left hand side of the `=`), and the content of the field (right hand side of `=`) ?

Reply:
Sorry, I'm not sure to follow. I think I much prefer the former example, not the latter. I think the way let-bindings are formatted is totally fine, and is the right way (consider the multi-line property differently for both the RHS and the body of the let).

My issue is rather with record fields, I think. Why:

```
{
   enum = {
     foo,
     bar,
   },
}
```

isn't reformated to

```nickel
{
   enum = 
   {
     foo,
     bar,
   },
}
```

? It seems everywhere else, mult-iline records get their first opening braces put on a fresh line, but not for record fields. At first sight, it might seem that a record field is considered single line because its definition up to `=` is on single line (for example, I imagine that if it were `enum | Foo | Bar | Baz =`, we would keep the metadata on one line). But I'm just guessing there. Is that right?

Reply:
Hmm... This proposal is way too coarse. You end up with things on the same line that have no place being there :disappointed:

As a compromise, I have come up with [quite complicated] rules that add the line break and indentation block in every case _except_ for when followed by a record, list or parens. Now we have this:

```nickel
# single/normal single/normal
let x = f x in f x

# single/normal single/exceptional
let x = f x in { x = x }

# single/normal multi/normal
let x = f x in
f x

# single/normal multi/exceptional
let x = f x in {
  x = x
}

# single/exceptional single/normal
let x = { x = 1 } in f x

# single/exceptional single/exceptional
let x = { x = 1 } in { x = x }

# single/exceptional multi/normal
let x = { x = 1 } in
f x

# single/exceptional multi/exceptional
let x = { x = 1 } in {
  x = x
}

# multi/normal single/normal
let x =
  f x
in f x

# multi/normal single/exceptional
let x =
  f x
in { x = x }

# multi/normal multi/normal
let x =
  f x
in
f x

# multi/normal multi/exceptional
let x =
  f x
in {
  x = x
}

# multi/exceptional single/normal
let x = {
  x = 1
} in f x

# multi/exceptional single/exceptional
let x = {
  x = 1
} in { x = x }

# multi/exceptional multi/normal
let x = {
  x = 1
} in
f x

# multi/exceptional multi/exceptional
let x = {
  x = 1
} in {
  x = x
}
```

The rules aren't perfect yet, but I think this is closer to what you're looking for... Please let me know :pray:

Reply:
Sorry, @yannham; I didn't see your comment while I was writing mine.

I'm not sure this answers your question, but...

So, IIUC, your proposal is to normalise the record field value formatting, when the value is a multi-line record, rather than normalising the `let` expression when multi-line records are involved? (And by "records", I mean "records, lists, parens and enums".)

I think the reason why I didn't go in that direction is because:

1. I was adding an extra level of indentation, so you'd get:

   ```nickel
   {
      enum = 
        {
          foo,
          bar,
        },
   }
   ```

2. While the indentation isn't necessary, opinionatedly, it seems common to see the opening brace at the end of a line, rather than on a new line on its own.

Reply:
I see, thanks. The indentation argument is indeed a good point.

Then what bothers me is the difference of treatment between record fields and LHS of a let-binding. I don't even think it's related to records, I guess the same thing happens with the opening bracket of an array, right? Record fields would be formatted like

```
{
  foo = [
    1,
    2
  ]
}
```

but let-bindings would be:

```
let foo =
  [
    1,
    2,
  ]
in
...
```

Here is my main constraint:

**if you take a let binding, erase the heading `let ` and the trailing `\nin`, that gives you a record field definition (well, you have to add enclosing `{`, and a trailing comma, but you get what I mean). I think the rest of this operation should be formatted the same as if it were a record field from the beginning.**

Otherwise, everything else is fine (including the `in` of let-bindings introducing a newline or not), but this property is not verified:

```nickel
let enum =
  {
    foo,
    bar
  }
in enum

# do the erase operation
enum =
  {
    foo,
    bar
  }

# but a record field would formatted as
enum = {
  foo,
  bar
}
```

You say we don't want to introduce an additional indent level for records, so we should keep the `enum = {` with the brace on the same line. Fair point! Then, would that be possible to that to the LHS of a let-binding to (not introducing an additional level even if it's multi-line), giving something similar to one of your suggestion, but not totally equivalent (the difference is that in the following, I didn't change the `in ...` introducing a newline before the RHS or not):

```
{
  single-single = let x = [ 1, 2 ] in x,

  single-multi = let x = [ 1, 2 ] in fun a =>
    x,

  multi-single = let x = [
    1,
    2
  ]
  in x,

  multi-multi = let x = [
    1,
    2
  ]
  in fun a =>
    x
}
```

But, in those examples, I would expect to the `in` to be on a line alone though. I don't know if it's related. So something like:

```
{
  single-single = let x = [ 1, 2 ] in x,

  single-multi = let x = [ 1, 2 ] in fun a =>
    x,

  multi-single = let x = [
    1,
    2
  ]
  in
  x,

  multi-multi = let x = [
    1,
    2
  ]
  in
  fun a =>
     x
}
```

Technically, I imagine it would amount to say that the only property affected by a let-binding being multi-line is to put the `in` on its own single line?

Reply:
In fact, I'm not sure anymore. There are also other situations when not putting the LHS of the let-binding makes it less nice. I think the current approach is already fine, and the path forward isn't totally clearcut. I will open an issue to discuss this on the Topiary repo directly, and gather more input.

In the meantime the current version is totally reasonable, so let's not block on this. Thanks for bearing with me and explaining this through!